powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ActivationServiceBehavior.java

```diff
@@ -1701,52 +1715,61 @@ public UpdateActivationOtpResponse updateActivationOtp(UpdateActivationOtpReques
-        // Check whether this is validation in the different step of activation. If yes, then the confirmation
-        // OTP must not be provided.
-        if (expectedStage != currentStage) {
-            if (confirmationOtp != null) {
-                logger.info("Activation OTP is not expected, but is provided: {}", activationId);
+            // Validate activation OTP is absent when it should not be present.
+            if (expectedStage != currentStage) {
+                if (StringUtils.hasText(confirmationOtp)) {
+                    logger.info("Activation OTP should not be present for activation ID: {}, stage: {}", activationId, currentStage);
+                    // Rollback is not required, database is not used for writing yet.
+                    throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_ACTIVATION_OTP);
+                }
+                // No validation required for this stage, the current stage is different from expected stage
+                return;
+            }
+        } else {
+            // Validation of commit phase:
+            // - if no OTP is specified for an activation, no validation is required
+            // - check each stage to make sure OTP is available when required
+            // - in case the commit is done in the next stage, skip validation
+            if (expectedOtpHash == null) {
+                return;
+            }
+            if (currentStage == ActivationOtpValidation.ON_KEY_EXCHANGE && commitPhase == CommitPhase.ON_KEY_EXCHANGE && !StringUtils.hasText(confirmationOtp)) {
+                logger.info("Activation OTP is missing for activation ID: {}, stage: {}", activationId, currentStage);
                 // Rollback is not required, database is not used for writing yet.
                 throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_ACTIVATION_OTP);
             }
-            return;
-        }
-        // We're in the right step, so the confirmation OTP must be provided.
-        if (confirmationOtp == null) {
-            logger.info("Activation OTP is expected, but is missing: {}", activationId);
-            // Rollback is not required, database is not used for writing yet.
-            throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_ACTIVATION_OTP);
+            if (currentStage == ActivationOtpValidation.ON_COMMIT && commitPhase == CommitPhase.ON_COMMIT && !StringUtils.hasText(confirmationOtp)) {
+                logger.info("Activation OTP is missing for activation ID: {}, stage: {}", activationId, currentStage);
+                // Rollback is not required, database is not used for writing yet.
+                throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_ACTIVATION_OTP);
+            }
+            if (currentStage == ActivationOtpValidation.ON_KEY_EXCHANGE && commitPhase == CommitPhase.ON_COMMIT) {
+                // Validation is done after key exchange, skip it for now
```

Comment:
We can actually do an analogic check if somebody provides `confirmationOtp` in the on key exchange and the expected phase is on commit. But it's not critical

Reply:
Yes, I can add this case.